File: remotedsh-contract/kotlin/VectorsTest.kt
// G2 门禁（docs/05 §2）：Kotlin 侧消费与 JS/Go 同一份共享向量。
// 位于 remotedsh-contract/kotlin；remotedsh-android 的 core-contract 模块以源码方式引入本目录（03 §3）。
// 注：android.util.Base64 在 JVM 单测中不可用，TurnCreds.credential 的单测由 androidTest 承担；
//     本文件覆盖 Frames / Hello / Fingerprint / QR 解析（QrPair.parse 用 android.net.Uri，同样归 androidTest）。
package dsh.mobile.contract

import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.nio.file.Files
import java.nio.file.Paths
import java.util.Base64

class VectorsTest {
    private fun vector(name: String): JSONObject = JSONObject(
        Files.readString(Paths.get("..", "vectors", name)),
    )

    // ---------- hello（契约 §4.2） ----------
    @Test
    fun helloCanonicalBytes() {
        val v = vector("hello-signature.json")
        val cases = v.getJSONArray("cases")
        for (i in 0 until cases.length()) {
            val c = cases.getJSONObject(i)
            if (!c.has("expected_msg_hex")) continue
            val got = Hello.canonicalBytes(c.getLong("ts"), c.getString("nonce"))
            val want = hex(c.getString("expected_msg_hex"))
            assertTrue(c.getString("name"), got.contentEquals(want))
        }
    }

    @Test
    fun helloVerifyVectors() {
        val v = vector("hello-signature.json")
        val priv = Hello.privateKeyFromPkcs8Der(
            pemBody(v.getJSONObject("_meta").getString("private_key_pkcs8_pem")),
        )
        val cases = v.getJSONArray("cases")
        // 向量公钥取自正例自带 pub（SPKI DER → base64url）；Kotlin 自产签名用同钥自验
        val pubDer = Base64.getUrlDecoder().decode(cases.getJSONObject(0).getString("pub"))
        for (i in 0 until cases.length()) {
            val c = cases.getJSONObject(i)
            val pub = if (c.getString("name") == "wrong pub key") {
                Base64.getUrlDecoder().decode(c.getString("pub"))
            } else pubDer
            val got = Hello.verify(Hello.publicKeyFromSpkiDer(pub), c.getLong("ts"), c.getString("nonce"), c.getString("expected_sig_base64url"))
            assertEquals(c.getString("name"), c.getBoolean("expect_verify"), got)
        }
        // Kotlin 自产签名（随机化 ECDSA）：自验证通过，且为 DER ≤72B
        val sig = Hello.sign(priv, 1770000000L, "nonce-from-kotlin")
        assertTrue(Hello.verify(Hello.publicKeyFromSpkiDer(pubDer), 1770000000L, "nonce-from-kotlin", sig))
        val der = Base64.getUrlDecoder().decode(sig)
        assertTrue(der[0] == 0x30.toByte() && der.size <= 72)
    }

    // ---------- 帧向量（契约 §3.1–§3.3） ----------
    @Test
    fun framesVectors() {
        val v = vector("frames.json")
        val types = v.getJSONObject("types")
        val cases = v.getJSONArray("cases")
        for (i in 0 until cases.length()) {
            val c = cases.getJSONObject(i)
            if (!c.has("frame_hex") || c.isNull("frame_hex")) {
                // construct 负例：len 超限走 encode 防线
                if (c.has("construct") && !c.isNull("construct")) {
                    val ctor = c.getJSONObject("construct")
                    try {
                        FrameCodec.encode(types.getInt("DATA"), ctor.getLong("stream_id"), ByteArray(ctor.getInt("payload_len")))
                        throw AssertionError("${c.getString("name")}: expected LEN_OVERFLOW")
                    } catch (e: FrameException) {
                        assertEquals("LEN_OVERFLOW", e.code)
                    }
                }
                continue
            }
            val bytes = hex(c.getString("frame_hex"))
            try {
                val f = FrameCodec.decode(bytes)
                if (c.has("decode_error") && !c.isNull("decode_error")) {
                    throw AssertionError("${c.getString("name")}: expected ${c.getString("decode_error")}, decoded fine")
                }
                val d = c.getJSONObject("decode")
                if (d.has("type_code") && !d.isNull("type_code")) {
                    assertEquals(c.getString("name"), d.getInt("type_code"), f.type)
                } else if (d.has("type")) {
                    assertEquals(c.getString("name"), types.getInt(d.getString("type")), f.type)
                }
                val wantFlags = if (d.has("flags_read_as") && !d.isNull("flags_read_as")) d.getInt("flags_read_as") else d.optInt("flags")
                assertEquals(c.getString("name"), wantFlags, f.flags)
                val wantStream = if (d.has("stream_id_read_as") && !d.isNull("stream_id_read_as")) d.getLong("stream_id_read_as") else d.optLong("stream_id")
                assertEquals(c.getString("name"), wantStream, f.streamId)
                if (d.has("payload_hex") && !d.isNull("payload_hex")) {
                    assertTrue(c.getString("name"), f.payload.contentEquals(hex(d.getString("payload_hex"))))
                }
                if (d.has("payload_utf8") && !d.isNull("payload_utf8")) {
                    assertEquals(c.getString("name"), d.getString("payload_utf8"), String(f.payload, Charsets.UTF_8))
                }
                if (d.has("payload_len") && !d.isNull("payload_len")) {
                    assertEquals(c.getString("name"), d.getInt("payload_len"), f.payload.size)
                }
            } catch (e: FrameException) {
                if (!(c.has("decode_error") && !c.isNull("decode_error"))) throw AssertionError("${c.getString("name")}: unexpected ${e.code}", e)
                assertEquals(c.getString("name"), c.getString("decode_error"), e.code)
            }
        }
    }

    // ---------- fingerprint（契约 §6.1） ----------
    @Test
    fun fingerprintVectors() {
        val v = vector("fingerprint-normalize.json")
        val cases = v.getJSONArray("cases")
        for (i in 0 until cases.length()) {
            val c = cases.getJSONObject(i)
            val lines = mutableListOf<String>()
            val arr = c.getJSONArray("offer_lines")
            for (j in 0 until arr.length()) lines.add(arr.getString(j))
            val got = Fingerprint.extract(lines)
            if (c.isNull("expect")) assertNull(c.getString("name"), got)
            else assertEquals(c.getString("name"), c.getString("expect"), got)
        }
    }

    private fun hex(s: String): ByteArray {
        val out = ByteArray(s.length / 2)
        for (i in out.indices) out[i] = s.substring(i * 2, i * 2 + 2).toInt(16).toByte()
        return out
    }

    private fun pemBody(pem: String): ByteArray =
        Base64.getMimeDecoder().decode(pem.replace(Regex("-----(BEGIN|END) PRIVATE KEY-----"), "").trim())
}

File: remotedsh-contract/go/vectors_test.go
// G2 门禁（docs/05 §2）：Go 侧消费与 JS/Kotlin 同一份共享向量。
// 运行：cd go && go test ./...
package contract

import (
	"bytes"
	"crypto/ecdsa"
	"crypto/hmac"
	"crypto/sha1"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"encoding/pem"
	"os"
	"path/filepath"
	"testing"
	"time"
)

const vectorsDir = "../vectors"

func loadVector(t *testing.T, name string) []byte {
	t.Helper()
	b, err := os.ReadFile(filepath.Join(vectorsDir, name))
	if err != nil {
		t.Fatalf("load %s: %v", name, err)
	}
	return b
}

// ---------- hello 签名向量（契约 §4.2） ----------

func TestHelloCanonicalBytes(t *testing.T) {
	var v struct {
		Cases []struct {
			Name            string `json:"name"`
			Ts              int64  `json:"ts"`
			Nonce           string `json:"nonce"`
			ExpectedMsgHex  string `json:"expected_msg_hex"`
			ExpectVerify    bool   `json:"expect_verify"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(loadVector(t, "hello-signature.json"), &v); err != nil {
		t.Fatal(err)
	}
	for _, c := range v.Cases {
		if c.ExpectedMsgHex == "" {
			continue
		}
		got := CanonicalBytes(c.Ts, c.Nonce)
		want, _ := hexDecode(c.ExpectedMsgHex)
		if !bytes.Equal(got, want) {
			t.Errorf("%s: canonical bytes mismatch", c.Name)
		}
	}
}

func TestHelloVerifyVectors(t *testing.T) {
	var v struct {
		Meta struct {
			PrivateKeyPkcs8Pem string `json:"private_key_pkcs8_pem"`
		} `json:"_meta"`
		Cases []struct {
			Name             string `json:"name"`
			Ts               int64  `json:"ts"`
			Nonce            string `json:"nonce"`
			Pub              string `json:"pub"`
			ExpectedMsgHex   string `json:"expected_msg_hex"`
			ExpectedSigB64u  string `json:"expected_sig_base64url"`
			ExpectVerify     bool   `json:"expect_verify"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(loadVector(t, "hello-signature.json"), &v); err != nil {
		t.Fatal(err)
	}
	// 向量密钥：PKCS8 PEM → ecdsa 私钥（Go 标准库随机化 ECDSA：只验证向量签名，不复现 RFC 6979）
	block, _ := pem.Decode([]byte(v.Meta.PrivateKeyPkcs8Pem))
	if block == nil {
		t.Fatal("bad test key PEM")
	}
	keyAny, err := x509.ParsePKCS8PrivateKey(block.Bytes)
	if err != nil {
		t.Fatal(err)
	}
	priv := keyAny.(*ecdsa.PrivateKey)
	pubDER, _ := x509.MarshalPKIXPublicKey(&priv.PublicKey)

	for _, c := range v.Cases {
		pub := pubDER
		if c.Name == "wrong pub key" {
			pub, _ = base64.RawURLEncoding.DecodeString(c.Pub)
		}
		got := VerifyHello(pub, c.Ts, c.Nonce, c.ExpectedSigB64u)
		if got != c.ExpectVerify {
			t.Errorf("%s: verify = %v, want %v", c.Name, got, c.ExpectVerify)
		}
	}
	// Go 自产签名：DER（SignASN1）、自验证通过
	sig, err := SignHello(priv, 1770000000, "nonce-from-go")
	if err != nil {
		t.Fatal(err)
	}
	if !VerifyHello(pubDER, 1770000000, "nonce-from-go", sig) {
		t.Fatal("Go sign/verify roundtrip failed")
	}
	sigDER, _ := base64.RawURLEncoding.DecodeString(sig)
	if len(sigDER) == 0 || sigDER[0] != 0x30 || len(sigDER) > 72 {
		t.Fatalf("sig not DER/≤72B: %d bytes", len(sigDER))
	}
	// DeviceID = sha256(SPKI DER) → base64url
	sum := sha256.Sum256(pubDER)
	if DeviceID(pubDER) != base64.RawURLEncoding.EncodeToString(sum[:]) {
		t.Fatal("DeviceID mismatch")
	}
}

func TestHelloProtoVer(t *testing.T) {
	if ProtoVer != "1.0.5" {
		t.Fatalf("ProtoVer = %q (契约 §7)", ProtoVer)
	}
}

// ---------- 帧向量（契约 §3.1/§3.2/§3.3） ----------

func TestFramesVectors(t *testing.T) {
	var v struct {
		Types map[string]int `json:"types"`
		Cases []struct {
			Name         string `json:"name"`
			FrameHex     *string `json:"frame_hex"`
			Construct    *struct {
				Type       string `json:"type"`
				StreamID   uint32 `json:"stream_id"`
				PayloadLen int    `json:"payload_len"`
			} `json:"construct"`
			DecodeError string `json:"decode_error"`
			Decode      *struct {
				Ver             int    `json:"ver"`
				Type            string `json:"type"`
				TypeCode        *int   `json:"type_code"`
				Flags           int    `json:"flags"`
				FlagsReadAs     *int   `json:"flags_read_as"`
				StreamID        uint32 `json:"stream_id"`
				StreamIDReadAs  *uint32 `json:"stream_id_read_as"`
				PayloadHex      *string `json:"payload_hex"`
				PayloadUtf8     *string `json:"payload_utf8"`
				Credit          *uint32 `json:"credit"`
				PayloadLen      *int    `json:"payload_len"`
			} `json:"decode"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(loadVector(t, "frames.json"), &v); err != nil {
		t.Fatal(err)
	}
	for _, c := range v.Cases {
		// 解码方向
		if c.FrameHex != nil {
			b, _ := hexDecode(*c.FrameHex)
			f, err := DecodeFrame(b)
			if c.DecodeError != "" {
				if err == nil {
					t.Errorf("%s: expected decode error %s, got nil", c.Name, c.DecodeError)
				}
				continue
			}
			if err != nil {
				t.Errorf("%s: decode: %v", c.Name, err)
				continue
			}
			d := c.Decode
			if d == nil {
				continue
			}
			if d.TypeCode != nil {
				if int(f.Type) != *d.TypeCode {
					t.Errorf("%s: type = %#x want %#x", c.Name, f.Type, *d.TypeCode)
				}
			} else if code, ok := v.Types[d.Type]; ok && int(f.Type) != code {
				t.Errorf("%s: type = %#x want %#x", c.Name, f.Type, code)
			}
			wantFlags := d.Flags
			if d.FlagsReadAs != nil {
				wantFlags = *d.FlagsReadAs
			}
			if f.Flags != uint8(wantFlags) {
				t.Errorf("%s: flags = %d want %d (接收方忽略非 0 不得拒帧)", c.Name, f.Flags, wantFlags)
			}
			if d.StreamIDReadAs != nil {
				if f.StreamID != *d.StreamIDReadAs {
					t.Errorf("%s: stream_id = %d want %d", c.Name, f.StreamID, *d.StreamIDReadAs)
				}
			} else if f.StreamID != d.StreamID {
				t.Errorf("%s: stream_id = %d want %d", c.Name, f.StreamID, d.StreamID)
			}
			if d.PayloadHex != nil {
				want, _ := hexDecode(*d.PayloadHex)
				if !bytes.Equal(f.Payload, want) {
					t.Errorf("%s: payload mismatch", c.Name)
				}
			}
			if d.PayloadUtf8 != nil && string(f.Payload) != *d.PayloadUtf8 {
				t.Errorf("%s: payload utf8 mismatch", c.Name)
			}
			if d.Credit != nil {
				if binary.BigEndian.Uint32(f.Payload) != *d.Credit {
					t.Errorf("%s: credit mismatch", c.Name)
				}
			}
			if d.PayloadLen != nil && len(f.Payload) != *d.PayloadLen {
				t.Errorf("%s: payload len = %d want %d", c.Name, len(f.Payload), *d.PayloadLen)
			}
		}
		// 编码方向（跳过接收侧宽容性用例）
		if c.Decode != nil && c.DecodeError == "" && c.Decode.Type != "" && c.FrameHex != nil {
			code, ok := v.Types[c.Decode.Type]
			if !ok {
				continue
			}
			var payload []byte
			switch {
			case c.Decode.PayloadHex != nil:
				payload, _ = hexDecode(*c.Decode.PayloadHex)
			case c.Decode.PayloadUtf8 != nil:
				payload = []byte(*c.Decode.PayloadUtf8)
			case c.Decode.Credit != nil:
				payload = make([]byte, 4)
				binary.BigEndian.PutUint32(payload, *c.Decode.Credit)
			case c.Decode.PayloadLen != nil:
				payload = bytes.Repeat([]byte{'a'}, *c.Decode.PayloadLen)
			}
			enc, err := EncodeFrame(uint8(code), c.Decode.StreamID, payload)
			if err != nil {
				t.Errorf("%s: encode: %v", c.Name, err)
				continue
			}
			if !bytes.Equal(enc, mustHex(t, *c.FrameHex)) {
				t.Errorf("%s: encode bytes mismatch", c.Name)
			}
			if enc[2] != 0 {
				t.Errorf("%s: sender MUST write flags=0", c.Name)
			}
		}
		// 负例 construct（len 超限走 encode 防线）
		if c.Construct != nil {
			_, err := EncodeFrame(TypeData, c.Construct.StreamID, bytes.Repeat([]byte{'a'}, c.Construct.PayloadLen))
			if err == nil {
				t.Errorf("%s: expected encode overflow", c.Name)
			}
		}
	}
}

// ---------- fingerprint 归一向量（契约 §6.1） ----------

func TestFingerprintVectors(t *testing.T) {
	var v struct {
		Cases []struct {
			Name       string   `json:"name"`
			OfferLines []string `json:"offer_lines"`
			Expect     *string  `json:"expect"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(loadVector(t, "fingerprint-normalize.json"), &v); err != nil {
		t.Fatal(err)
	}
	for _, c := range v.Cases {
		got := ExtractPinnedFingerprint(c.OfferLines)
		want := ""
		if c.Expect != nil {
			want = *c.Expect
		}
		if got != want {
			t.Errorf("%s: got %q want %q", c.Name, got, want)
		}
	}
}

// ---------- QR 向量（契约 §2） ----------

func TestQRVectors(t *testing.T) {
	var v struct {
		Cases []struct {
			Name   string  `json:"name"`
			URI    string  `json:"uri"`
			Expect *struct {
				V  int     `json:"v"`
				S  string  `json:"s"`
				PT string  `json:"pt"`
				FP string  `json:"fp"`
				A  string  `json:"a"`
				T  *string `json:"t"`
			} `json:"expect"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(loadVector(t, "qr-pair.json"), &v); err != nil {
		t.Fatal(err)
	}
	for _, c := range v.Cases {
		got, err := ParsePairQR(c.URI)
		if c.Expect == nil {
			if err == nil {
				t.Errorf("%s: expected error", c.Name)
			}
			continue
		}
		if err != nil {
			t.Errorf("%s: %v", c.Name, err)
			continue
		}
		if got.S != c.Expect.S || got.PT != c.Expect.PT || got.FP != c.Expect.FP || got.A != c.Expect.A {
			t.Errorf("%s: fields mismatch: %+v", c.Name, got)
		}
	}
}

// ---------- TURN（契约 §5） ----------

func TestTurn(t *testing.T) {
	now := time.Unix(1770000000, 0)
	if got := TurnUsername("pair-1", 3600, now); got != "1770003600:pair-1" {
		t.Fatalf("username = %q", got)
	}
	cred := TurnCredential("secret-s", "1770003600:pair-1")
	mac := hmacSha1("secret-s", "1770003600:pair-1")
	if cred != mac {
		t.Fatalf("credential = %q want %q", cred, mac)
	}
	uris := TurnUris("bridge.example.com")
	if len(uris) != 3 || uris[2] != "turn:bridge.example.com:3478?transport=tcp" {
		t.Fatalf("uris = %v", uris)
	}
}

// ---------- helpers ----------

func hexDecode(s string) ([]byte, error) { return hex.DecodeString(s) }

func mustHex(t *testing.T, s string) []byte {
	t.Helper()
	b, err := hex.DecodeString(s)
	if err != nil {
		t.Fatal(err)
	}
	return b
}

func hmacSha1(secret, msg string) string {
	mac := hmac.New(sha1.New, []byte(secret))
	mac.Write([]byte(msg))
	return base64.StdEncoding.EncodeToString(mac.Sum(nil))
}
